README.md
# discord-web-mirror

Create RSS mirrors of Discord channels, complete with embed and attachment support.

- Create an RSS mirror of a Discord channel
- Attachments, embeds and metadata (excluding author data) all retained
- Retroactively fetches past channel messages

## Why?

`discord-web-mirror` was intended to mirror news-based Discord channels, a lot of community news is crowd-sourced and published in Discord channels.

## Setup

### Requirements
- Node.js v14.0+
- Full-time server
- Discord private token
  1. Discord doesn't like when you do this, so accept that this might result in an account ban.
  2. Open Discord in your browser.
  3. Open your browser's web tools and look in the *Network* tab.
  4. Find a request to the Discord API (try typing in the message box to trigger `typing`) and look at the request headers.
  5. Copy the token under the *Authorization* header.

### Guide
1. Install NPM dependencies with `npm install`.
2. Rename `config.yaml.example` to `config.yaml` and enter your Discord and web server info in the file.
3. Run the server with `node index.js`.

## How?

`discord-web-mirror` works by making interval requests to the private Discord API to pull recent messages from a channel and then saving those messages in a database. These archived messages are then formatted into an RSS feed at request.

### Infrastructure

1. Interval GET requests to the Discord API (`/api/v9/channels/${channelId}/messages`)
2. Saves all message data to the SQLite database
3. Saves all attachments and embed images to `static/`
4. Express.js serves the messages from the SQLite database in an RSS feed, alongside static hosting for the attachments (`/static`).

### Privacy

No author data is saved, not even randomized identifiers are attached to messages, it is impossible to determine who wrote a message from the message database.

CHANGELOG.md
# Changelog

## [Unreleased]

### Added



src/db.js
const crypto = require("crypto")

function setup(db) {
    db.prepare(`CREATE TABLE IF NOT EXISTS guild_metadata (
        id TEXT PRIMARY KEY,
        name TEXT NOT NULL,
        icon_url TEXT 
    )`).run()

    db.prepare(`CREATE TABLE IF NOT EXISTS channel_metadata (
        id TEXT PRIMARY KEY,
        guild_id TEXT NOT NULL,
        name TEXT NOT NULL,
        FOREIGN KEY(guild_id) REFERENCES guild_metadata(id)
    )`).run()

    db.prepare(`CREATE TABLE IF NOT EXISTS messages (
        id TEXT PRIMARY KEY,
        type INT NOT NULL,
        content TEXT,
        channel_id TEXT NOT NULL,
        timestamp INT NOT NULL,
        edited_timestamp INT,
        referenced_message_id TEXT,
        contains_attachments BOOLEAN,
        contains_embeds BOOLEAN,
        FOREIGN KEY(referenced_message_id) REFERENCES messages(id)
    )`).run()

    db.prepare(`CREATE TABLE IF NOT EXISTS message_attachments (
        id TEXT PRIMARY KEY,
        message_id TEXT NOT NULL,
        filename TEXT NOT NULL,
        size INT NOT NULL,
        url TEXT NOT NULL,
        proxy_url TEXT,
        content_type TEXT,
        FOREIGN KEY(message_id) REFERENCES messages(id)
    )`).run()

	db.prepare(`CREATE TABLE IF NOT EXISTS message_embeds (
		hash TEXT PRIMARY KEY,
		message_id TEXT,
		sort_index NUMBER NOT NULL,
		type TEXT NOT NULL,
		url TEXT,
		title TEXT,
		description TEXT,
		color INT,
		
		author_name TEXT,
		author_url TEXT,
		author_icon_url TEXT,
		author_proxy_icon_url TEXT,
		
		image_url TEXT,
		image_proxy_url TEXT,
		
		thumbnail_url TEXT,
		thumbnail_proxy_url TEXT,
		
		footer_text TEXT,
		footer_icon_url TEXT,
		footer_proxy_icon_url TEXT,
		
		FOREIGN KEY(message_id) REFERENCES messages(id)
	)`).run()
}
exports.setup = setup

function insertGuildMetadata(db, id, name, iconUrl) {
    db.prepare("UPDATE guild_metadata SET name=?, icon_url=? WHERE id=?").run(name, iconUrl, id)
    db.prepare("INSERT or IGNORE INTO guild_metadata (id, name, icon_url) VALUES (?, ?, ?)").run(id, name, iconUrl)
}
exports.insertGuildMetadata = insertGuildMetadata

function getGuildMetadata(db, id) {
    const guild = db.prepare("SELECT * FROM guild_metadata WHERE id=?").get(id)
    if (guild == null) {
        throw new Error("No guild with that id")
    }
    return {
        id: guild.id,
        name: guild.name,
        iconUrl: guild.icon_url
    }
}
exports.getGuildMetadata = getGuildMetadata

function insertChannelMetadata(db, id, guildId, name) {
    db.prepare("UPDATE channel_metadata SET guild_id=?, name=? WHERE id=?").run(guildId, name, id)
    db.prepare("INSERT or IGNORE INTO channel_metadata (id, guild_id, name) VALUES (?, ?, ?)").run(id, guildId, name)
}
exports.insertChannelMetadata = insertChannelMetadata

function getChannelMetadata(db, id) {
    const channel = db.prepare("SELECT * FROM channel_metadata WHERE id=?").get(id)
    if (channel == null) {
        return null
    }
    return {
        id: channel.id,
        guildId: channel.guild_id,
        name: channel.name,
    }
}
exports.getChannelMetadata = getChannelMetadata

function channelExists(db, id) {
	const query = db.prepare("SELECT id FROM channel_metadata WHERE id=?").get(id)
	return query != null
}
exports.channelExists = channelExists

function getRecentChannelMessage(db, channelId) {
	const query = db.prepare("SELECT id, timestamp, edited_timestamp FROM messages WHERE channel_id=? ORDER BY timestamp DESC, edited_timestamp DESC").get(channelId)
	if (query?.edited_timestamp) {
		query.editedTimestamp = query.edited_timestamp
		delete query.edited_timestamp
	}
	return query
}
exports.getRecentChannelMessage = getRecentChannelMessage

function insertDiscordMessages(db, messages) {
    function epochTimestamp(dateString) {
        const date = new Date(dateString)
        return Math.floor(date.getTime() / 1000)
    }

    function insertMessage(message, referencedMessageId = null) {
        let editedTimestamp = null
        if (message.edited_timestamp != null) {
            editedTimestamp = epochTimestamp(message.edited_timestamp)
        }

		const containsAttachments = message.attachments.length > 0 ? 1 : 0
		const containsEmbeds = message.embeds.length > 0 ? 1 : 0

        db.prepare(`UPDATE messages SET content=?, edited_timestamp=?, contains_attachments=?, contains_embeds=? WHERE id=?`).run(message.content, editedTimestamp, containsAttachments, containsEmbeds, message.id)
        db.prepare(`INSERT or IGNORE INTO messages (id, type, content, channel_id, timestamp, edited_timestamp, referenced_message_id, contains_attachments, contains_embeds) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`)
        .run(message.id, message.type, message.content, message.channel_id, epochTimestamp(message.timestamp), editedTimestamp, referencedMessageId, containsAttachments, containsEmbeds)
        
        for (const embedIndex in message.embeds) {
        	const embed = message.embeds[embedIndex]
        	
        	const hash = crypto.createHash("md5").update(`${message.id}${embedIndex}`).digest("hex")
        	db.prepare(`INSERT or IGNORE INTO message_embeds (
        		hash, message_id, sort_index, type, url, title, description, color,
        		author_name, author_url, author_icon_url, author_proxy_icon_url,
        		image_url, image_proxy_url,
        		thumbnail_url, thumbnail_proxy_url,
        		footer_text, footer_icon_url, footer_proxy_icon_url) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`)
        	.run(
        		hash, message.id, embedIndex, embed.type, embed.url, embed.title, embed.description, embed.color,
        		embed.author?.name, embed.author?.url, embed.author?.icon_url, embed.author?.proxy_icon_url,
        		embed.image?.url, embed.image?.proxy_url,
        		embed.thumbnail?.url, embed.thumbnail?.proxy_url,
        		embed.footer?.text, embed.footer?.icon_url, embed.footer?.proxy_icon_url)
        }
        
        for (const attachment of message.attachments) {
        	db.prepare(`INSERT or IGNORE INTO message_attachments (id, message_id, filename, size, url, proxy_url, content_type) VALUES (?, ?, ?, ?, ?, ?, ?)`)
        	.run(attachment.id, message.id, attachment.filename, attachment.size, attachment.url, attachment.proxy_url, attachment.content_type)
        }
    }

    const insertManyMessages = db.transaction(messages => {
        for (const message of messages) {
            insertMessage(message)
            if (message.message_reference && message.referenced_message) {
                insertMessage(message.referenced_message)
                db.prepare("UPDATE messages SET referenced_message_id=? WHERE id=?").run(message.message_reference.message_id, message.id)
            }
        }
    })
    insertManyMessages(messages)
}
exports.insertDiscordMessages = insertDiscordMessages

function collectPrefixedKeys(obj, prefix) {
	var newObj = {}
	for (const key in obj) {
		if (key.startsWith(prefix)) {
			const newKey = key.slice(prefix.length)
			if (obj[key])
				newObj[newKey] = obj[key]
			delete obj[key]
		}
	}
	return Object.keys(newObj).length <= 0 ? null : newObj
}
function trimNullValues(obj) {
	for (const key in obj) {
		if (obj[key] == null) {
			delete obj[key]
		}
	}
}

function getDiscordMessages(db, channelId, count = 50, afterMessageId = undefined) {
	const messageCount = Math.min(Math.max(count, 0), 50)
	
	let rows = []
	if (afterMessageId) {
		rows = db.prepare("SELECT * FROM messages WHERE channel_id = ? AND id < ? ORDER BY timestamp DESC LIMIT ?").all(channelId, afterMessageId, messageCount)
	} else {
		rows = db.prepare("SELECT * FROM messages WHERE channel_id = ? ORDER BY timestamp DESC LIMIT ?").all(channelId, messageCount)
	}
	
	for (row of rows) {
		// pull in attachments
		if (row.contains_attachments) {
			const attachmentRows = db.prepare("SELECT * FROM message_attachments WHERE message_id = ?").all(row.id)
			for (const attachment of attachmentRows) {
				trimNullValues(attachment)
			}
			row.attachments = attachmentRows
		}
		delete row.contains_attachments
		
		// pull in embeds
		if (row.contains_embeds) {
			const embedRows = db.prepare("SELECT * FROM message_embeds WHERE message_id = ? ORDER BY sort_index ASC").all(row.id)
			var completeEmbeds = []
			for (const embedIndex in embedRows) {
				const embed = embedRows[embedIndex]
			
				embed.author = collectPrefixedKeys(embed, "author_")
				const image = collectPrefixedKeys(embed, "image_")
				embed.images = image != null ? [ image ] : []
				embed.thumbnail = collectPrefixedKeys(embed, "thumbnail_")
				embed.footer = collectPrefixedKeys(embed, "footer_")
				
				if (embed.color) {
					const hexColor = embed.color
					embed.color = "#" + hexColor.toString(16)
				}
				
				if (embed.author || embed.thumbnail || embed.footer || embed.title || embed.description) {
					completeEmbeds.push(embed)
				} else if (image != null && completeEmbeds.length > 0) {
					completeEmbeds.at(-1).images.push(image)
					continue
				}
				
				trimNullValues(embed)
				delete embed.sort_index
			}
			row.embeds = completeEmbeds
		}
		delete row.contains_embeds
		
		trimNullValues(row)
	}
	return rows
}
exports.getDiscordMessages = getDiscordMessages


routes/app.js
const crypto = require("crypto")

const express = require("express")
const router = express.Router()

const sanitizeHtml = require("sanitize-html")
const marked = require("marked")
const { htmlTrimBlankLines } = require("html-trim-blank-lines")

const dbHelper = require("../src/db")

function customDateFormatGMT(date) {
    const gmtDate = new Date(date.toLocaleString("en-US", { timeZone: "GMT" }))
    var pm = gmtDate.getHours() > 12
    var hours = gmtDate.getHours() % 12
    return `${months[gmtDate.getMonth()]} ${gmtDate.getDate()}, ${gmtDate.getFullYear()} at ${hours}:${String(gmtDate.getMinutes()).padStart(2, "0")} ${pm ? "pm" : "am"} GMT`
}

function generateUrl(config, url, type) {
    const urlHash = crypto.createHash("md5").update(url).digest("hex")
    const urlObj = new URL(url)
    return config.server.url + `/static/${type}/${urlHash}/` + urlObj.pathname.split("/").at(-1)
}

// wrapper for html trim blank lines
router.use((req, res, next) => {
	function process(callback) {
		return function(err, html) {
			res.send(htmlTrimBlankLines(html))
		}
	}

	res.oldRender = res.render
	res.render = function(view, options, callback) {
		res.oldRender(view, options, process(callback))
	}

	return next()
})

// const months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
const months = ["Jan", "Feb", "March", "April", "May", "June", "July", "Aug", "Sept", "Oct", "Nov", "Dec"]

router.get("/channels", (req, res) => {
	const db = req.app.get("db")
	const config = req.app.get("config")
	
	const channels = config.discord.channels.reduce((filtered, channelId) => {
		const channel = dbHelper.getChannelMetadata(db, channelId)
		if (channel != null) {
			const guild = dbHelper.getGuildMetadata(db, channel.guildId)
			channel.guild = guild

            const recentMessage = dbHelper.getRecentChannelMessage(db, channelId)
            const recentMessageTimestamp = recentMessage?.editedTimestamp || recentMessage?.timestamp
            if (recentMessageTimestamp) {
                channel.lastUpdated = recentMessageTimestamp
                const dateString = customDateFormatGMT(new Date(recentMessageTimestamp * 1000))
                channel.humanLastUpdated = dateString
            } else {
                channel.lastUpdated = 0
                channel.humanLastUpdated = "No messages"
            }

			filtered.push(channel)
		}
		return filtered
	}, []).sort((a, b) => b.lastUpdated - a.lastUpdated)

    const channelsLastUpdated = channels.map(channel => {
        return { id: channel.id, lastUpdated: channel.lastUpdated }
    })

    const channelsCount = `${channels.length} channel${channels.length == 0 || channels.length > 1 ? "s" : ""}`
   
    const metadata = {
        app: config.app,
        humanChannelsCount: channelsCount,
        channelsLastUpdated: channelsLastUpdated
    }
	
	res.render("pages/channels_list.ejs", { channels: channels, config: config, metadata: metadata })
})

function fetchChannelPosts(db, config, channelId, postLimit) {
    const posts = dbHelper.getDiscordMessages(db, channelId, postLimit)
	posts.forEach(post => {
        const basename = config.server.url + "/static"
		post.html_content = sanitizeHtml(marked.parse(post.content))
        post.iso_timestamp = new Date(post.timestamp * 1000).toISOString()
        post.human_timestamp = customDateFormatGMT(new Date(post.timestamp * 1000))
        post.edited = post.editedTimestamp != null
        post.attachments?.forEach(postAttachment => {
            var url = postAttachment.proxy_url
            const sourceUrl = new URL(postAttachment.url)
            // fixing a bug where the proxied version of some cdn attachments fail to GET
            if (sourceUrl.hostname == "cdn.discordapp.com") {
                url = postAttachment.url
            }
            postAttachment.resolved_url_local = generateUrl(config, url, "attachments")
        })
        post.embeds?.forEach(embed => {
            if (embed.description) {
                embed.html_description = sanitizeHtml(marked.parse(embed.description))
            }
            embed.images?.forEach(embedImage => {
                embedImage.proxy_url_local = generateUrl(config, embedImage.proxy_url, "external")
            })

            if (embed.thumbnail) {
                embed.thumbnail.proxy_url_local = generateUrl(config, embed.thumbnail.proxy_url, "external")
            }
        })
	})
    return posts
}

router.get("/channels/:channel_id", (req, res) => {
    const db = req.app.get("db")
    const config = req.app.get("config")

    const channelId = req.params.channel_id
    if (!config.discord.channels.includes(channelId)) {
        res.sendStatus(404)
        return
    }

    let channelMetadata
    try {
        channelMetadata = dbHelper.getChannelMetadata(db, channelId)
    } catch (e) {
        console.log(e)
        res.sendStatus(404)
        return
    }
    
    let guildMetadata
    try {
        guildMetadata = dbHelper.getGuildMetadata(db, channelMetadata.guildId)
    } catch (e) {
        console.log(e)
        res.sendStatus(404)
        return
    }

    const posts = fetchChannelPosts(db, config, channelId, 50)

    const metadata = {
        id: channelMetadata.id,
        name: channelMetadata.name,
        guild: {
            id: guildMetadata.id,
            name: guildMetadata.name,
            icon_url_local: generateUrl(config, guildMetadata.iconUrl, "metadata")
        }
    } 

    res.render("pages/channel.ejs", { posts: posts, metadata: metadata, config: config })
})

router.get("/channels/:channel_id/feed", (req, res) => {
    const db = req.app.get("db")
    const config = req.app.get("config")

    const channelId = req.params.channel_id
    if (!config.discord.channels.includes(channelId)) {
        res.sendStatus(404)
        return
    }

    let channelMetadata
    try {
        channelMetadata = dbHelper.getChannelMetadata(db, channelId)
    } catch (e) {
        console.log(e)
        res.sendStatus(404)
        return
    }
    
    let guildMetadata
    try {
        guildMetadata = dbHelper.getGuildMetadata(db, channelMetadata.guildId)
    } catch (e) {
        console.log(e)
        res.sendStatus(404)
        return
    }

    const server = {
        link: `${config.server.url}/${channelId}/feed` 
    }

    const recentMessage = dbHelper.getRecentChannelMessage(db, channelId)
    const recentMessageTimestamp = recentMessage?.editedTimestamp || recentMessage?.timestamp
    const lastUpdated = recentMessageTimestamp != null ? new Date(parseInt(recentMessageTimestamp) * 1000) : new Date()

    const posts = fetchChannelPosts(db, config, channelId, 50)

    const metadata = {
        id: channelMetadata.id,
        name: channelMetadata.name,
        last_updated_iso: lastUpdated.toISOString(),
        guild: {
            id: guildMetadata.id,
            name: guildMetadata.name,
            icon_url_local: generateUrl(config, guildMetadata.iconUrl, "metadata")
        }
    } 

    res.setHeader("Content-Type", "text/xml; charset=utf-8")
	res.render("pages/atom.ejs", { server: server, metadata: metadata, posts: posts })
})

router.get("/", (req, res) => {
    const config = req.app.get("config")
	res.redirect(`${config.server.url}/channels`)
})

module.exports = router


src/config.js
const { parse, stringify } = require("yaml")
const path = require("path")
const fs = require("fs")
const { log } = require("./log")

function exit(msg) {
    log("error", "config", msg)
    process.exit(1)
}

function readConfig() {
    const configPath = path.join(__dirname, "..", "config.yaml")
    const data = fs.readFileSync(configPath, "utf8")
    const config = parse(data)

    if (!config) { exit("invalid config, must be valid yaml") }

    if (!config.discord) { config.discord = {} }
    if (!config.discord?.token) { exit("missing discord token") }
    if (!config.discord?.channels) { exit("missing discord channels") }
    if (!config.discord?.approx_interval) { config.discord.approx_interval = 600 }

    if (!config.server) { config.server = {} }
    if (!config.server.url) { exit("missing server url") }
    if (config.server.log == undefined) { config.server.log = false }
    if (config.server.enabled == undefined) { config.server.enabled = true }

    return config
}
exports.readConfig = readConfig

routes/atom.js
const express = require("express")
const router = express.Router()

const sanitizeHtml = require("sanitize-html")
const marked = require("marked")

const dbHelper = require("../src/db")

router.get("/", (req, res) => {
    const db = req.app.get("db")

    const posts = dbHelper.getDiscordMessages(db, process.env.DISCORD_CHANNEL_ID, 50)
	posts.forEach(post => {
		post.html_content = sanitizeHtml(marked.parse(post.content))
        post.iso_timestamp = new Date(post.timestamp * 1000).toISOString()
        post.attachments?.forEach(postAttachment => {
            const attachmentUrl = new URL(postAttachment.proxy_url)
            postAttachment.proxy_url_local = "http://localhost:3000/static" + attachmentUrl.pathname
        })
        post.embeds?.forEach(embed => {
            if (embed.description) {
                embed.html_description = sanitizeHtml(marked.parse(embed.description))
            }
            embed.images?.forEach(embedImage => {
                const imageUrl = new URL(embedImage.proxy_url)
                embedImage.proxy_url_local = "http://localhost:3000/static" + imageUrl.pathname
            })

            if (embed.thumbnail) {
                const thumbnailUrl = new URL(embed.thumbnail.proxy_url)
                embed.thumbnail.proxy_url_local = "http://localhost:3000/static" + thumbnailUrl.pathname
            }
        })
	})
	res.render("pages/atom.ejs", { posts: posts })
})

module.exports = router

src/discord.js
const axios = require("axios")
const fs = require("fs")
const path = require("path")
const util = require("util")
const stream = require("stream")
const pipeline = util.promisify(stream.pipeline)
const dbHelper = require("./db")
const { DiscordError } = require("./error")
const { log } = require("./log")
const crypto = require("crypto")

// should return the server name and url to guild picture
async function getGuildMetadata(token, guildId) {
    const response = await axios(`https://discord.com/api/v9/guilds/${guildId}`, {
        headers: {
            "authorization": token,
            "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36"
        }
    })
    if (response.status != 200) {
        throw new Error(`Discord responded with status code ${response.status}`)
    }
    const guildIcon = `https://cdn.discordapp.com/icons/${guildId}/${response.data.icon}.jpeg`
    return {
        id: response.data.id,
        name: response.data.name,
        iconUrl: guildIcon
    }
}
exports.getGuildMetadata = getGuildMetadata

async function getChannelMetadata(token, channelId) {
    const response = await axios(`https://discord.com/api/v9/channels/${channelId}`, {
        headers: {
            "authorization": token,
            "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36"
        }
    })
    if (response.status != 200) {
        throw new Error(`Discord responded with status code ${response.status}`)
    }
    return {
        id: response.data.id,
        name: response.data.name,
        guildId: response.data.guild_id
    }
}
exports.getChannelMetadata = getChannelMetadata

async function getChannelMessages(token, channelId) {
    const response = await axios(`https://discord.com/api/v9/channels/${channelId}/messages?limit=50`, {
        headers: {
            "authorization": token,
            "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36"
        }
    })
    
    if (response.status == 200) {
        return response.data
    } else {
        throw new Error(`Discord responded with status code ${response.status}`)
    }
}
exports.getChannelMessages = getChannelMessages

async function saveMessageAttachment(rawUrl, type = "attachments", override = false) {
	const url = new URL(rawUrl)
    const urlHash = crypto.createHash("md5").update(rawUrl).digest("hex")
	const filename = url.pathname.split("/").at(-1)
    const dirPath = path.join("static", type, urlHash)
	const fullPath = path.join(dirPath, filename)
	
	if (override == false && fs.existsSync(fullPath)) {
		return
	}
	
	await fs.promises.mkdir(dirPath, { recursive: true })
	
	log("discord dl", "info", `downloading ${rawUrl} to ${fullPath}`)
	const response = await axios.get(rawUrl, {
		responseType: "stream"
	})
	await pipeline(response.data, fs.createWriteStream(fullPath))
	log("discord dl", "info", `download complete of ${rawUrl}`)
}
exports.saveMessageAttachment = saveMessageAttachment

async function archiveChannelMessages(db, discordToken, discordChannelId) {
	log("discord", "info", `archiving messages in channel ${discordChannelId}`)
	let msgs
	try {
		msgs = await getChannelMessages(discordToken, discordChannelId)
		log("discord", "info", `retrieved ${msgs.length} messages from channel ${discordChannelId}`)
	} catch (err) {
		// wrap discord api provided errors
		
		if (err instanceof axios.AxiosError && err.response?.status != 200) {
			throw new DiscordError(err.response.data.message)	
		} else {
			throw err
		}
// 		throw err
	}
	
	// save the msgs to the database
	dbHelper.insertDiscordMessages(db, msgs)
	
	log("discord", "info", `downloading attachments from channel ${discordChannelId}`)
	// download each attachment from each message
	
	try {
		for (const msg of msgs) {
			if (msg.attachments.length > 0) {
				for (const attachment of msg.attachments) {
					var url = attachment.proxy_url

					const sourceUrl = new URL(attachment.url)
					// fixing a bug where the proxied version of some cdn attachments fail to GET
					if (sourceUrl.hostname == "cdn.discordapp.com") {
						url = attachment.url
					}

					await saveMessageAttachment(url, "attachments")
				}
			}
			if ((msg.embeds?.length || 0) > 0) {
				for (const embed of msg.embeds) {
					if (embed.author?.proxy_icon_url)
						await saveMessageAttachment(embed.author.proxy_icon_url, "external")
					if (embed.image?.proxy_url)
						await saveMessageAttachment(embed.image.proxy_url, "external")
					if (embed.thumbnail?.proxy_url)
						await saveMessageAttachment(embed.thumbnail.proxy_url, "external")
				}
			}
		}
	} catch (err) {
		console.log(err)
	}
	
	
	
// 	try {
// 		await Promise.all(downloads)
// 	} catch (err) {
// 		if (err.response) {
// 			log("discord dl", "warn", `download failed - ${err.response.status}`)
// 		}
// 		log("discord dl", "warn", `downloaded failed`)
// 	}
	
	log("discord", "info", `downloaded all attachments from channel ${discordChannelId}`)
}
exports.archiveChannelMessages = archiveChannelMessages


index.js
require("dotenv").config()

const express = require("express")
const app = express()

const dbHelper = require("./src/db")
const configHelper = require("./src/config")
const discord = require("./src/discord")
const { log } = require("./src/log")

const config = configHelper.readConfig()
config.app = {
	name: "discord-web-mirror",
	version: "v0.2.0 beta"
}

const db = require("better-sqlite3")("discord-web-mirror.sqlite3")
dbHelper.setup(db)

if (config.server.log) {
	const morgan = require("morgan")
	app.use(morgan("combined"))
}

app.use("/", require("./routes/app"))
app.use("/static", express.static("static"))

app.set("view engine", "ejs")
app.set("config", config)
app.set("db", db)

// start the web server if enabled

if (config.server.enabled) {
	const port = config.server.port || process.env.PORT || 3000
	app.listen(port, () => {
		log("web server", "info", `web server hosted on port ${port}`)
	})
} else {
	log("web server", "info", "web server disabled in config")
}

// channel metadata and archive channel messages

async function collectChannelMetadata(channels) {
    var guildIds = []
    for (const channelId of config.discord.channels) {
        const channelMetadata = await discord.getChannelMetadata(config.discord.token, channelId)
        if (!guildIds.includes(channelMetadata.guildId)) {
            guildIds.push(channelMetadata.guildId)
            const guildMetadata = await discord.getGuildMetadata(config.discord.token, channelMetadata.guildId)
            dbHelper.insertGuildMetadata(db, guildMetadata.id, guildMetadata.name, guildMetadata.iconUrl)
            await discord.saveMessageAttachment(guildMetadata.iconUrl, "metadata", true)
        }
        dbHelper.insertChannelMetadata(db, channelMetadata.id, channelMetadata.guildId, channelMetadata.name)
    }
}

async function discordArchiveInterval(channelId) {
	await discord.archiveChannelMessages(db, config.discord.token, channelId)
	const min = Math.max(config.discord.approx_interval - 120, 0)
	const max = config.discord.approx_interval + 120
	const intervalSeconds = Math.floor(Math.random() * (max - min + 1)) + min
	log("discord", "info", `archiving channel ${channelId} again in ${intervalSeconds} seconds (${intervalSeconds / 60} minutes)`)
	setTimeout(() => {
		discordArchiveInterval(channelId)
	}, intervalSeconds * 1000)
}

collectChannelMetadata(config.discord.channels).then(_ => {
    for (const channelId of config.discord.channels) {
        discordArchiveInterval(channelId)
    }
})

